OpenRA.Mods.Common/Widgets/Logic/IrcLogic.cs

```diff
@@ -0,0 +1,266 @@
+
+			connectBG = widget.Get("IRC_CONNECT_BG");
+			ircContainer = widget.Get("IRC_CONTAINER");
+
+			var disconnectButton = widget.Get<ButtonWidget>("DISCONNECT_BUTTON");
+			disconnectButton.IsDisabled = () => client == null;
+			disconnectButton.OnClick = Disconnect;
+
+			MaybeShowConnectPanel();
+		}
+
+		void MaybeShowConnectPanel()
+		{
+			if (client != null && client.IsConnected)
+			{
+				ircContainer.Visible = true;
+				connectBG.Visible = false;
+				return;
+			}
+
+			if (Game.Settings.Irc.ConnectAutomatically)
+			{
+				ircContainer.Visible = true;
+				connectBG.Visible = false;
+				Connect();
+				return;
+			}
+
+			ircContainer.Visible = false;
+			connectBG.Visible = true;
+
+			var connectAutomaticallyCheckBox = connectBG.Get<CheckboxWidget>("CONNECT_AUTOMATICALLY_CHECKBOX");
+			connectAutomaticallyCheckBox.IsChecked = () => Game.Settings.Irc.ConnectAutomatically;
+			connectAutomaticallyCheckBox.OnClick = () => Game.Settings.Irc.ConnectAutomatically ^= true;
+
+			var connectButton = connectBG.Get<ButtonWidget>("CONNECT_BUTTON");
+			connectButton.OnClick = () =>
+			{
+				ircContainer.Visible = true;
+				connectBG.Visible = false;
+
+				Game.Settings.Irc.ConnectAutomatically = connectAutomaticallyCheckBox.IsChecked();
+				Game.Settings.Save();
+				Connect();
+			};
+		}
+
+		void Connect()
+		{
+			client = new IrcClient();
+			client.Encoding = System.Text.Encoding.UTF8;
+			client.SendDelay = 100;
+			client.ActiveChannelSyncing = true;
+
+			client.OnConnected += OnConnected;
+			client.OnError += OnError;
+			client.OnRawMessage += OnRawMessage;
+			client.OnJoin += OnJoin;
+			client.OnChannelActiveSynced += OnChannelActiveSynced;
+
+			try
+			{
+				AddChatLine("Connecting to {0}...".F(Game.Settings.Irc.Hostname));
+				client.Connect(Game.Settings.Irc.Hostname, Game.Settings.Irc.Port);
+			}
+			catch (Exception e)
+			{
+				AddChatLine("Connection error: {0}".F(e.Message));
+				Log.Write("irc", e.ToString());
+			}
+
+			listenThread = new Thread(new ThreadStart(Listen));
+			listenThread.Start();
+		}
+
+		public void Listen()
+		{
+			client.Listen();
+		}
+
+		Widget MakeLabelWidget(LabelWidget template, string item)
+		{
+			var widget = (LabelWidget)template.Clone();
+			var font = Game.Renderer.Fonts[widget.Font];
+			item = WidgetUtils.WrapText(item, widget.Bounds.Width, font);
+			widget.Bounds.Height = font.Measure(item).Y;
+			widget.GetText = () => item;
+			return widget;
+		}
+
+		void AddChatLine(string text)
+		{
+			Log.Write("irc", text);
+
+			var scrolledToBottom = historyPanel.ScrolledToBottom;
+
+			var newChild = MakeLabelWidget(historyTemplate, text);
+			historyPanel.AddChild(newChild);
+
+			if (scrolledToBottom)
+				historyPanel.ScrollToBottom(smooth: true);
+		}
+
+		bool EnterPressed()
+		{
+			if (!inputBox.Text.Any())
+				return true;
+
+			var text = inputBox.Text;
+			inputBox.Text = "";
+
+			if (text.StartsWith("/"))
+				AddChatLine("Unknown command");
```

Comment:
If we don't want users to perform commands, why not print something like
`You're not allowed to perform commands/this command!` out?